File: Data Structures & Algorithms/time-based-key-value-store/submission-0.py
class TimeMap:

    def __init__(self):
        self.d=defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.d[key].append([value,timestamp])

    def get(self, key: str, timestamp: int) -> str:
        l=0
        if key not in self.d:
            return ""
        r=len(self.d[key])-1
        ans=""
        k=0
        while(l<=r):
            mid=l+(r-l)//2
            if(self.d[key][mid][1]<=timestamp):
                l=mid+1
                if self.d[key][mid][1]>=k:
                    k=self.d[key][mid][1]
                    ans=self.d[key][mid][0]
            else:
                r=mid-1
        return ans


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: Data Structures & Algorithms/lru-cache/submission-0.cpp
class LRUCache {
public:
    struct Node{
        int value;
        int key;
        Node* next;
        Node* prev;
    };
    int cap;
    LRUCache(int capacity) {
        cap=capacity;
    }
    Node* head=nullptr;
    Node* tail=nullptr;
    unordered_map<int,Node*> mp;
    int count=0;

    void del(Node* n){
        Node* t;
        if(n==head){
            t=head;
            head=head->next;
            if(head)
                head->prev=nullptr;
            else
                tail=nullptr;
        }
        else if(n==tail){
            t=tail;
            tail=tail->prev;
            tail->next=nullptr;
        }
        else{
            t=n;
            t->next->prev=t->prev;
            t->prev->next=t->next;
        }
        mp.erase(t->key);
        count--;
    }
    void add(int key,int value){
        Node* n = new Node();
        n->key=key;
        n->value=value;
        mp[n->key]=n;
        count++;
        if(tail!=nullptr){
            tail->next=n;
            n->prev=tail;}
        else{
            head=n;
        }
        tail=n;
    }
    int get(int key) {
        if(mp.find(key)!=mp.end()){
            put(key,mp[key]->value);
            return mp[key]->value;}
        else
            return -1;
    }
    
    void put(int key, int value) {
         
        if(mp.find(key)!=mp.end()){
            del(mp[key]);
        }
        else if(count==cap){
            Node* t=head;
            del(t);
        }
        add(key,value);
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */